pytest_operator/plugin.py

```diff
@@ -420,53 +570,135 @@ async def run(self, *cmd, cwd=None, check=False, fail_msg=None):
 
     _run = run  # backward compatibility alias
 
-    async def juju(self, *args):
+    async def juju(self, *args, **kwargs):
         """Runs a Juju CLI command.
 
         Useful for cases where python-libjuju sees things differently than the Juju CLI.
         Will set `JUJU_MODEL`, so manually passing in `-m model-name` is unnecessary.
         """
 
-        return await self.run("juju", *args)
+        return await self.run("juju", *args, **kwargs)
+
+    async def _add_model(self, controller_name, cloud_name, model_name, **kwargs):
+        """
+        Creates a model used by the test framework which would normally be destroyed
+        after the tests are run in the module.
+        """
+        controller = self._controller
+        if not controller:
+            controller = Controller()
+            await controller.connect(controller_name)
+        if not cloud_name:
+            # if not provided, try the default cloud name
+            cloud_name = self._init_cloud_name
+        if not cloud_name:
+            # if not provided, use the controller's default cloud
+            cloud_name = await controller.get_cloud()
+        model_full_name = f"{controller_name}:{model_name}"
+        log.info(f"Adding model {model_full_name} on cloud {cloud_name}")
+
+        model_keep = False
+        model = await controller.add_model(model_name, cloud_name, **kwargs)
+        # NB: This call to `juju models` is needed because libjuju's
+        # `add_model` doesn't update the models.yaml cache that the Juju
+        # CLI depends on with the model's UUID, which the CLI requires to
+        # connect. Calling `juju models` beforehand forces the CLI to
+        # update the cache from the controller.
+        await self.juju("models")
+        state = ModelState(model, model_keep, controller_name, cloud_name, model_name)
```

Comment:
Is `ModelState.keep` ever used?  I guess in a test someone could make a test case that updates this?

Reply:
It's used [here](https://github.com/charmed-kubernetes/pytest-operator/pull/69/files#diff-e67fdec84b020bfd0fa3e0554fc6687c8167c33f2d23c0bae53449816f8b59e2R511-R517)

during the `forget_model` method, if `self.keep_model` is true, then the model won't be deleted -- only disconnected.

Reply:
yeah I see where we read it, but I wasn't sure how someone sets `keep=True` so that their model will be kept.

Reply:
yes, this needs to be addressed as there's a gap here that isn't covered.  I'm looking to make this clear.  The goal is ultimately flexibility. 

each model managed by `ops_test` should be either `ephemeral` and created by the test infrastructure or `stateful` where it's state is maintained before/after the tests are completed. 


I think this just needs a better API in `track_model`to reflect the two paths


```python
ops_test.track_model("alias")  # make a new model with any juju name and destroy it when the tests are over
ops_test.track_model("alias", keep=True)  # make a new model with any juju name and don't destroy it when the tests are over
ops_test.track_model("alias", model_name="bob")  # make a new model known to juju as "bob" and destroy it when the tests are over
ops_test.track_model("alias", model_name="bob", keep=True)  # make a new model known to juju as  "bob" and don't destroy it when the tests are over

```

Reply:
@ca-scribner i believe [this](https://github.com/charmed-kubernetes/pytest-operator/pull/69/commits/a68e0739f02a5885a5c2c2bfb742317fa59d6ba8) commit will clear up your questions